src/BitBucketOnPrem.ts

```diff
@@ -28,28 +36,79 @@ export interface BitbucketOnPrem {
     branch?: string;
     fileTree?: string[];
     filePath?: string;
+    treeSize?: number;
+}
+
+export interface BitbucketOnPremRepoProps {
+    projectKey: string;
+    repoName: string;
+    commit: string;
+    searchTerm?: string;
+    maxResults?: number;
 }
 
 export interface BitBucketOnPremInput {
     args: BitbucketOnPrem;
 }
 
+export interface BitbucketOnPremTreeInput {
+    args: BitbucketOnPremRepoProps;
+}
+
+const getRepoId = ({projectKey, repoName, commit}: {projectKey: string, repoName: string, commit: string}) => {
+    return `${projectKey}::${repoName}::${commit}`;
+};
+
+const getRepoFromId = (id: string): BitbucketOnPremRepoProps => {
+    const parts = _.split(id, "::");
+    return {
+        projectKey: parts[0],
+        repoName: parts[1],
+        commit: parts[2]
+    };
+};
+
 // fetchNoCache fetches a resource without loading/saving cache and also avoids using cookies.
-// Otherwise we get inconsistent results from bitbucket API with different tokens
+// Otherwise, we get inconsistent results from bitbucket API with different tokens
 const fetchNoCache = (requestInfo: RequestInfo, requestInit: RequestInit) => {
     requestInit = requestInit || {};
     if (!requestInit?.cache) {
-        requestInit.cache = 'no-store';
+        requestInit.cache = "no-store";
     }
     if (!requestInit?.credentials) {
-        requestInit.credentials = 'omit';
+        requestInit.credentials = "omit";
     }
-    return fetch(requestInfo.toString().replace('scm/', '').replace('scm%2F', ''), requestInit);
+    return fetch(requestInfo.toString().replace("scm/", "").replace("scm%2F", ""), requestInit);
+};
+
+/**
+ * Fetch 5 pages of the tree in parallel
+ * If a page is beyond the amount of files the repo has, it will return an empty array (the behaviour of the bitbucket api)
+ * @param fileTreeUrl
+ * @param accessToken
+ * @param start
+ * @param limit
+ * @return Array of file arrays, each array is from a different page of the tree api call
+ */
+const fetchTreeParallel =
+    async ({fileTreeUrl, accessToken, start, limit}: {fileTreeUrl: string; accessToken: string; start: number; limit: number}):
+        Promise<string[][]> => {
+    const requests = _.map([0, 1, 2, 3, 4], async reqIndex => {
```

Comment:
There is no api limit from BB ?

Reply:
@hopemanryan When you exceed the limit, it returns an empty array, which is checked afterwards

Reply:
as in the api rate limiter ?